src/Transform/ONNX/ONNXDimAnalysis.hpp

```diff
@@ -0,0 +1,72 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===-------- ONNXDimAnalysis.hpp - ONNX Dimension Analysis ---------------===//
+//
+// Copyright 2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file implements an analysis on unknown dimensions in ONNX ops.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "mlir/IR/BuiltinTypes.h"
+
+namespace onnx_mlir {
+
+class DimAnalysis {
+public:
+  using DimT = std::pair<mlir::Value, uint64_t>;
+  using DimSetT = llvm::SmallDenseSet<DimT, 4>;
+  using DimSetMapT = llvm::SmallDenseMap<uint64_t, DimSetT, 4>;
+
+public:
+  /// Create a new analysis for specific values.
+  DimAnalysis(llvm::ArrayRef<mlir::Value> vals);
+
+  /// Create a new analysis for all values in a module.
+  DimAnalysis(mlir::ModuleOp op);
+
+  /// Analyzes the relationship amongs unknown dimensions.
+  void analyze();
+
+  /// Returns the grouping result of unknown dimensions.
+  DimSetMapT getGroupingResult() const { return dimSetMap; }
+
+  /// Test if two unknown dimensions are the same or not.
```

Comment:
what are `lhs` and `rhs`? Are they value of a shaped type kind? So `tensorOrMemref`? If so, a more explicit name would facilitate the usage of the interface.